### PR TITLE
Enable final stack post-processing for transformer for correct prenorm behavior

### PR DESCRIPTION
**Pull request by emjotde**
### Description

This PR enables final post-processing of a full transformer stack for correct prenorm behavior.
See issues: #715 and #699,

List of changes:
- Add final post-processing in encoder and decoder if requested with `--transformer-postprocess-top`. Can take combinations of `d, n, a`. Using `a` will add a skip connection from the bottom of the stack.
- Add `--task transformer-base-prenorm` and `--task transformer-big-prenorm` which correspond to `--task transformer-base --transformer-preprocess n --transformer-postprocess da --transformer-postprocess-top n`.

Added dependencies: none

### Checklist

- [x] I have tested the code manually
- [x] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

### CODE REVIEW

**Comment by XapaJIaMnu** [comment]
I have started testing on my models. I will report the results later.
With this pull request, I noticed that the GPU only build is broken due to faiss (idk when it was added)
`make VERBOSE=1` yields:
```
/usr/bin/c++   -std=c++11 -pthread -Wl,--no-as-needed -fPIC -Wno-unused-result -Wno-unknown-warning-option  -march=native  -msse2 -msse3 -msse4.1 -msse4.2 -mavx -mavx2 -mavx512f -DUSE_SENTENCEPIECE -DCUDA_FOUND -DUSE_NCCL -O3 -m64 -funroll-loops -g -rdynamic   CMakeFiles/marian_train.dir/command/marian_main.cpp.o  -o ../marian ../libmarian.a -ldl /usr/local/cuda/lib64/libcurand.so /usr/local/cuda/lib64/libcusparse.so /usr/lib/x86_64-linux-gnu/libcublas.so ../local/lib/libnccl_static.a /usr/lib/libtcmalloc_minimal.so -ldl /usr/local/cuda/lib64/libcurand.so /usr/local/cuda/lib64/libcusparse.so /usr/lib/x86_64-linux-gnu/libcublas.so ../local/lib/libnccl_static.a /usr/lib/libtcmalloc_minimal.so -lpthread ../libmarian.a libmarian_cuda.a 3rd_party/sentencepiece/src/libsentencepiece_train.so.0.0.0 -ldl /usr/local/cuda/lib64/libcurand.so /usr/local/cuda/lib64/libcusparse.so /usr/lib/x86_64-linux-gnu/libcublas.so ../local/lib/libnccl_static.a /usr/lib/libtcmalloc_minimal.so -lpthread 3rd_party/sentencepiece/src/libsentencepiece.so.0.0.0 /usr/lib/x86_64-linux-gnu/libprotobuf.so /usr/local/cuda/lib64/libcurand.so /usr/local/cuda/lib64/libcusparse.so /usr/lib/x86_64-linux-gnu/libcublas.so /usr/lib/libtcmalloc_minimal.so /usr/local/cuda/lib64/libcudart_static.a -ldl -lpthread /usr/lib/x86_64-linux-gnu/librt.so -Wl,-rpath,/usr/local/cuda/lib64:/marian-dev/build/src/3rd_party/sentencepiece/src  -lcblas
../libmarian.a(VectorTransform.cpp.o): In function `(anonymous namespace)::eig(unsigned long, double*, double*, int)':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:429: undefined reference to `dsyev_'
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:434: undefined reference to `dsyev_'
../libmarian.a(VectorTransform.cpp.o): In function `faiss::LinearTransform::transform_transpose(long, float const*, float*) const':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:292: undefined reference to `sgemm_'
../libmarian.a(VectorTransform.cpp.o): In function `matrix_qr(int, int, float*)':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:99: undefined reference to `sgeqrf_'
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:104: undefined reference to `sgeqrf_'
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:107: undefined reference to `sorgqr_'
../libmarian.a(VectorTransform.cpp.o): In function `faiss::LinearTransform::apply_noalloc(long, float const*, float*) const':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:268: undefined reference to `sgemm_'
../libmarian.a(VectorTransform.cpp.o): In function `faiss::LinearTransform::set_is_orthonormal()':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:320: undefined reference to `sgemm_'
../libmarian.a(VectorTransform.cpp.o): In function `faiss::PCAMatrix::prepare_Ab()':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:712: undefined reference to `sgemm_'
../libmarian.a(VectorTransform.cpp.o): In function `faiss::PCAMatrix::train(long, float const*)':
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:519: undefined reference to `ssyrk_'
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:560: undefined reference to `ssyrk_'
/marian-dev/src/3rd_party/faiss/VectorTransform.cpp:600: undefined reference to `sgemm_'
collect2: error: ld returned 1 exit status
```
Maybe we need GPU-only CI test?

**Comment by emjotde** [comment]
Looks like my fault, I moved a couple of ifdefs around recently and didn't catch that.

**Comment by XapaJIaMnu** [comment]
Initial testing suggests that for my (artificial) low resource scenario, this configuration (transformer-base-prenorm) trains faster and better than the transformer-base.

**Comment by emjotde** [comment]
And no NANs with TensorCores etc? Neat.

**Comment by XapaJIaMnu** [comment]
Yeah, that was on 2080tis, no NaNs. Should I try fp16?

**Comment by emjotde** [comment]
No need for now. I only recommended the fp16 branch because it has NaN-handling stuff. If it's not even happening then there is no other gain.

**Comment by emjotde** [comment]
The failing compilation, was that happening without a MKL or other BLAS installation?

**Comment by XapaJIaMnu** [comment]
I used `-DCOMPILE_CPU=OFF` to completely disable CPU. Both BLAS and MKL were present otherwise.

**Comment by emjotde** [comment]
OK, this is ready for review now.

**Comment by emjotde** [comment]
@XapaJIaMnu was able to reproduce that and already fixed in master. One missing ifdef.

File: src/models/transformer.h [hunk 1]
```diff
@@ -558,6 +558,8 @@ class EncoderTransformer : public Transformer<EncoderBase> {
     auto layer     = transposeTimeBatch(batchEmbeddings); // [beam depth=1, batch size, max length, vector dim]
     auto layerMask = transposeTimeBatch(batchMask);       // [beam depth=1, batch size, max length, vector dim=1]
 
+    auto prevLayer = layer; // keep handle to untransformed embeddeddings, potentially used for a final skip connection
```
Comment (snukky) [inline src/models/transformer.h:561]: A typo in "embeddings"

File: src/models/transformer.h [hunk 2]
```diff
@@ -706,6 +714,8 @@ class DecoderTransformer : public Transformer<DecoderBase> {
     // reorganize batch and timestep
     auto query = transposeTimeBatch(scaledEmbeddings); // [-4: beam depth=1, -3: batch size, -2: max length, -1: vector dim]
 
+    auto prevQuery = query;
```
Comment (snukky) [inline src/models/transformer.h:717]: Does the variable deserve a comment as `prevLayer` above?

File: src/models/transformer.h [hunk 3]
```diff
@@ -580,6 +582,12 @@ class EncoderTransformer : public Transformer<EncoderBase> {
       checkpoint(layer); // sets a manually specified checkpoint if gradient checkpointing is enabled, does nothing otherwise.
     }
 
+    // This allows to run a final layernorm operation after going through the transformer layer stack.
+    // By default the operations are empty, but with prenorm (--transformer-preprocess n --transformer-postprocess da) 
+    // we should normalize here. Can also be used to add a skip connection from the very bottom if requested.
```
Comment (snukky) [inline src/models/transformer.h:587]: I found that "we should normalize here" leaves the comment a little unclear. Is it a recommendation or todo, and who is "we" referring to, a user or developer? Consider making it clearer.

File: src/models/transformer.h [hunk 4]
```diff
@@ -841,6 +851,12 @@ class DecoderTransformer : public Transformer<DecoderBase> {
       checkpoint(query);
     }
 
+    // This allows to run a final layernorm operation after going through the transformer layer stack.
+    // By default the operations are empty, but with prenorm (--transformer-preprocess n --transformer-postprocess da) 
+    // we should normalize here. Can also be used to add a skip connection from the very bottom if requested.
```
Comment (snukky) [inline src/models/transformer.h:856]: Consider making the comment clearer, see my comment above.

**Review by snukky — APPROVED** [review]
A typo and minor suggestions.

File: src/models/transformer.h (hunk 3 above)
Comment (emjotde) [reply src/models/transformer.h:587]: Hm, I don't read it like that. It is actually a recommendation and not a requirement. All of this is quite fuzzy.

**Review by XapaJIaMnu — APPROVED** [review]
I think we should leave in the comments references to the initial transformer implementation and/or the two papers I mentioned in the issue.

**Comment by emjotde** [comment]
@XapaJIaMnu I'll add references to the implementations. The papers are about other things?
BTW. For large data sets prenorm seems to be noticeably worse for me in terms of BLEU (similar perpelexity though) than the default postnorm, so careful.

**Pull request merged**